### PR TITLE
removing retries from s3 blob store. Close #22845

### LINKED ISSUE #22845: Investigate removing retries from s3 blob store
The s3 client has a maxRetries setting, which is uses internally for retrying throttled requests. The setting is also read by the s3 blob store, and a while{try/catch} is used along with retries (but without backoff). It appears this outer retry mechanism may have existed for throttling before the sdk client did that, or also for eventual consistency. For the latter, this should no longer be a problem, as currently s3 provides read-after-write consistency for new PUTs (and afaik we should not be overwriting blobs). This issue is to discuss removing the retries from the blob store (but not the client).

### PR DESCRIPTION
**Pull request by lewischen1123**
removing retries from s3 blob store. Close #22845.

### CODE REVIEW

**Review by abeyad — CHANGES_REQUESTED** [review]
@lewischen1123 thank you for contributing to Elasticsearch.  

I think the goal with this issue was to keep the max retry setting on the AWS S3 client (so that we can still configure the number of retries we set on the S3 client), but remove the retry logic from the `S3BlobStore` class.  That would mean getting rid of the `numberOfRetries` in the `S3BlobStore`, fixing the `S3BlobStore#readBlob()` method so retries aren't used (since they are taken care of by the S3 client), and removing the retries logic from the `S3OutputStream`.

Lastly, we will need you to kindly sign the CLA.

File: plugins/repository-s3/src/main/java/org/elasticsearch/cloud/aws/blobstore/DefaultS3OutputStream.java [hunk 1]
```diff
@@ -107,12 +107,12 @@ public void flush(byte[] bytes, int off, int len, boolean closing) throws IOExce
     private void upload(byte[] bytes, int off, int len) throws IOException {
         try (ByteArrayInputStream is = new ByteArrayInputStream(bytes, off, len)) {
             int retry = 0;
-            while (retry <= getNumberOfRetries()) {
+            while (retry <= 0) {
```
Comment (abeyad) [inline plugins/repository-s3/src/main/java/org/elasticsearch/cloud/aws/blobstore/DefaultS3OutputStream.java]: we should get rid of the retry logic here entirely, there is no need for the while loop when we aren't retrying.

File: plugins/repository-s3/src/main/java/org/elasticsearch/cloud/aws/blobstore/DefaultS3OutputStream.java [hunk 2]
```diff
@@ -160,15 +160,15 @@ protected void doUpload(S3BlobStore blobStore, String bucketName, String blobNam
 
     private void initializeMultipart() {
         int retry = 0;
-        while ((retry <= getNumberOfRetries()) && (multipartId == null)) {
+        while ((retry <= 0) && (multipartId == null)) {
```
Comment (abeyad) [inline plugins/repository-s3/src/main/java/org/elasticsearch/cloud/aws/blobstore/DefaultS3OutputStream.java]: same here, all retry logic should be removed

File: plugins/repository-s3/src/main/java/org/elasticsearch/cloud/aws/blobstore/DefaultS3OutputStream.java [hunk 3]
```diff
@@ -160,15 +160,15 @@ protected void doUpload(S3BlobStore blobStore, String bucketName, String blobNam
...
             try {
                 multipartId = doInitialize(getBlobStore(), getBucketName(), getBlobName(), isServerSideEncryption());
                 if (multipartId != null) {
                     multipartChunks = 1;
                     multiparts = new ArrayList<>();
                 }
             } catch (AmazonClientException e) {
-                if (getBlobStore().shouldRetry(e) && retry < getNumberOfRetries()) {
+                if (getBlobStore().shouldRetry(e) && retry < 0) {
```
Comment (abeyad) [inline plugins/repository-s3/src/main/java/org/elasticsearch/cloud/aws/blobstore/DefaultS3OutputStream.java]: same here with removing retry logic

File: plugins/repository-s3/src/main/java/org/elasticsearch/cloud/aws/blobstore/DefaultS3OutputStream.java [hunk 4]
```diff
@@ -194,14 +194,14 @@ protected String doInitialize(S3BlobStore blobStore, String bucketName, String b
     private void uploadMultipart(byte[] bytes, int off, int len, boolean lastPart) throws IOException {
         try (ByteArrayInputStream is = new ByteArrayInputStream(bytes, off, len)) {
             int retry = 0;
-            while (retry <= getNumberOfRetries()) {
+            while (retry <= 0) {
```
Comment (abeyad) [inline plugins/repository-s3/src/main/java/org/elasticsearch/cloud/aws/blobstore/DefaultS3OutputStream.java]: same here with removing retry logic

File: plugins/repository-s3/src/main/java/org/elasticsearch/cloud/aws/blobstore/DefaultS3OutputStream.java [hunk 5]
```diff
@@ -194,14 +194,14 @@ protected String doInitialize(S3BlobStore blobStore, String bucketName, String b
...
                 try {
                     PartETag partETag = doUploadMultipart(getBlobStore(), getBucketName(), getBlobName(), multipartId, is, len, lastPart);
                     multiparts.add(partETag);
                     multipartChunks++;
                     return;
                 } catch (AmazonClientException e) {
-                    if (getBlobStore().shouldRetry(e) && retry < getNumberOfRetries()) {
+                    if (getBlobStore().shouldRetry(e) && retry < 0) {
```
Comment (abeyad) [inline plugins/repository-s3/src/main/java/org/elasticsearch/cloud/aws/blobstore/DefaultS3OutputStream.java]: same here with removing retry logic

File: plugins/repository-s3/src/main/java/org/elasticsearch/cloud/aws/blobstore/DefaultS3OutputStream.java [hunk 6]
```diff
@@ -231,13 +231,13 @@ protected PartETag doUploadMultipart(S3BlobStore blobStore, String bucketName, S
 
     private void completeMultipart() {
         int retry = 0;
-        while (retry <= getNumberOfRetries()) {
+        while (retry <= 0) {
```
Comment (abeyad) [inline plugins/repository-s3/src/main/java/org/elasticsearch/cloud/aws/blobstore/DefaultS3OutputStream.java]: same here with removing retry logic

File: plugins/repository-s3/src/main/java/org/elasticsearch/cloud/aws/blobstore/DefaultS3OutputStream.java [hunk 7]
```diff
@@ -231,13 +231,13 @@ protected PartETag doUploadMultipart(S3BlobStore blobStore, String bucketName, S
...
             try {
                 doCompleteMultipart(getBlobStore(), getBucketName(), getBlobName(), multipartId, multiparts);
                 multipartId = null;
                 return;
             } catch (AmazonClientException e) {
-                if (getBlobStore().shouldRetry(e) && retry < getNumberOfRetries()) {
+                if (getBlobStore().shouldRetry(e) && retry < 0) {
```
Comment (abeyad) [inline plugins/repository-s3/src/main/java/org/elasticsearch/cloud/aws/blobstore/DefaultS3OutputStream.java]: same here with removing retry logic

File: plugins/repository-s3/src/main/java/org/elasticsearch/cloud/aws/blobstore/DefaultS3OutputStream.java [hunk 8]
```diff
@@ -73,7 +73,7 @@
     private List<PartETag> multiparts;
 
     public DefaultS3OutputStream(S3BlobStore blobStore, String bucketName, String blobName, int bufferSizeInBytes, int numberOfRetries, boolean serverSideEncryption) {
```
Comment (abeyad) [inline plugins/repository-s3/src/main/java/org/elasticsearch/cloud/aws/blobstore/DefaultS3OutputStream.java]: number of retries should be removed as a constructor arg

**Review by abeyad — CHANGES_REQUESTED** [review]
Also, the `S3BlobStore` class does not need any retries code anymore (don't need the class field, or anything else related to it).  The only place where the configured number of max retries needs to remain is in `S3Repository`, which passes that value to the `AwsS3Service`.

**Comment by rjernst** [comment]
@lewischen1123 Are you still interested in finishing up this PR?

**Pull request closed without merging**